File: app/src/main/java/com/aslansari/hypocoin/ui/CurrencyFragment.java
package com.aslansari.hypocoin.ui;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.aslansari.hypocoin.R;
import com.aslansari.hypocoin.app.HypoCoinApp;
import com.aslansari.hypocoin.repository.CoinService;
import com.aslansari.hypocoin.repository.model.Currency;
import com.aslansari.hypocoin.ui.adapters.CurrencyRecyclerAdapter;
import com.aslansari.hypocoin.viewmodel.CoinViewModel;

import java.util.List;

import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.subscribers.DisposableSubscriber;
import timber.log.Timber;

/**
 * A simple {@link Fragment} subclass.
 * Use the {@link CurrencyFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class CurrencyFragment extends Fragment {

    private ProgressBar progressBar;
    private CoinViewModel coinViewModel;
    private CurrencyRecyclerAdapter currencyRecyclerAdapter;
    private CompositeDisposable disposables;
    private CoinService coinService;
    private boolean isBoundCoinService;
    private ServiceConnection serviceConnection;

    public CurrencyFragment() {
        // Required empty public constructor
    }

    public static CurrencyFragment newInstance() {
        CurrencyFragment fragment = new CurrencyFragment();
        Bundle args = new Bundle();
        // add params if necessary
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            // get params if exists
        }
        coinViewModel = ((HypoCoinApp) getActivity().getApplication()).appContainer.coinViewModel;
        disposables = new CompositeDisposable();

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_currency, container, false);
        RecyclerView recyclerView = view.findViewById(R.id.recyclerCoin);
        progressBar = view.findViewById(R.id.progressBar);
        currencyRecyclerAdapter = new CurrencyRecyclerAdapter();
        LinearLayoutManager layoutManager = new LinearLayoutManager(getActivity(),
                RecyclerView.VERTICAL, false);
        recyclerView.setLayoutManager(layoutManager);
        DividerItemDecoration dividerItemDecoration =
                new DividerItemDecoration(recyclerView.getContext(), layoutManager.getOrientation());
        recyclerView.addItemDecoration(dividerItemDecoration);
        recyclerView.setAdapter(currencyRecyclerAdapter);

        serviceConnection = new ServiceConnection() {
            @Override
            public void onServiceConnected(ComponentName name, IBinder service) {
                coinService = ((CoinService.LocalBinder) service).getService();

                disposables.add(coinService.getAsyncCurrencies()
                        .subscribeWith(new DisposableSubscriber<List<Currency>>() {
                            @Override
                            public void onNext(List<Currency> currencyList) {
                                currencyRecyclerAdapter.updateList(currencyList);
                            }

                            @Override
                            public void onError(Throwable t) {

                            }

                            @Override
                            public void onComplete() {

                            }
                        })
                );

                coinService.start();
            }

            @Override
            public void onServiceDisconnected(ComponentName name) {
                coinService = null;
            }
        };

        Context context = getActivity();
        if (context != null) {
            isBoundCoinService = context.bindService(new Intent(getActivity(), CoinService.class), serviceConnection, Context.BIND_AUTO_CREATE);
        }

        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        disposables.add(coinViewModel.getCurrencyList()
                .subscribeWith(new DisposableSubscriber<Currency>() {
                    @Override
                    protected void onStart() {
                        progressBar.setVisibility(View.VISIBLE);
                        super.onStart();
                    }

                    @Override
                    public void onNext(Currency currency) {
                        progressBar.setVisibility(View.GONE);
                        currencyRecyclerAdapter.add(currency);
                    }

                    @Override
                    public void onError(@NonNull Throwable e) {
                        Timber.e(e);
                    }

                    @Override
                    public void onComplete() {
                        Timber.d("onComplete");
                    }
                }));
    }

    @Override
    public void onStart() {
        super.onStart();
        if (coinService != null) {
            coinService.start();
        }
    }

    @Override
    public void onStop() {
        super.onStop();
        coinService.stop();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        disposables.dispose();
        currencyRecyclerAdapter = null;
        Context context = getActivity();
        if (context != null && isBoundCoinService) {
            context.unbindService(serviceConnection);
        }
    }
}

File: app/src/main/java/com/aslansari/hypocoin/ui/CurrencyFragment.kt
package com.aslansari.hypocoin.ui

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.aslansari.hypocoin.R
import com.aslansari.hypocoin.databinding.FragmentCurrencyBinding
import com.aslansari.hypocoin.repository.CoinService
import com.aslansari.hypocoin.repository.CoinService.LocalBinder
import com.aslansari.hypocoin.repository.model.Currency
import com.aslansari.hypocoin.ui.adapters.CurrencyRecyclerAdapter
import com.aslansari.hypocoin.ui.adapters.MarginItemDecorator
import com.aslansari.hypocoin.viewmodel.CoinViewModel
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.subscribers.DisposableSubscriber

/**
 * A simple [Fragment] subclass.
 * Use the [CurrencyFragment.newInstance] factory method to
 * create an instance of this fragment.
 */
class CurrencyFragment : BaseFragment() {
    private val coinViewModel: CoinViewModel by viewModels(factoryProducer = {
        viewModelCompositionRoot.viewModelFactory
    })
    private var currencyRecyclerAdapter: CurrencyRecyclerAdapter? = null
    private var disposables: CompositeDisposable? = null
    private var coinService: CoinService? = null
    private var isBoundCoinService = false
    private var serviceConnection: ServiceConnection? = null
    private var binding: FragmentCurrencyBinding? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            // get params if exists
        }
        disposables = CompositeDisposable()
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentCurrencyBinding.inflate(inflater, container, false)
        currencyRecyclerAdapter = CurrencyRecyclerAdapter()
        return binding?.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding?.apply {
            recyclerCoin.layoutManager =
                LinearLayoutManager(requireActivity(), RecyclerView.VERTICAL, false)
            val verticalMargin = resources.getDimensionPixelSize(R.dimen.currency_margin_vertical)
            val horizontalMargin = resources.getDimensionPixelSize(R.dimen.currency_margin_horizontal)
            recyclerCoin.addItemDecoration(MarginItemDecorator(verticalMargin, horizontalMargin))
            recyclerCoin.adapter = currencyRecyclerAdapter
            vm = coinViewModel
            lifecycleOwner = viewLifecycleOwner
        }
        coinViewModel.currencyList.observe(viewLifecycleOwner) { currencyList ->
            currencyRecyclerAdapter?.updateList(currencyList)
        }
        coinViewModel.getCurrencyList()
        // TODO: update ui with flowable instead of service 
        serviceConnection = object : ServiceConnection {
            override fun onServiceConnected(name: ComponentName, service: IBinder) {
                coinService = (service as LocalBinder).service
                disposables!!.add(coinService!!.asyncCurrencies
                    .subscribeWith(object : DisposableSubscriber<List<Currency?>?>() {
                        override fun onNext(currencyList: List<Currency?>?) {
                            // currencyRecyclerAdapter?.updateList(currencyList as List<Currency>)
                        }

                        override fun onError(t: Throwable) {}
                        override fun onComplete() {}
                    })
                )
                coinService!!.start()
            }

            override fun onServiceDisconnected(name: ComponentName) {
                coinService = null
            }
        }
        isBoundCoinService = requireActivity().bindService(Intent(activity, CoinService::class.java),
            serviceConnection as ServiceConnection,
            Context.BIND_AUTO_CREATE)
    }

    override fun onStart() {
        super.onStart()
        if (coinService != null) {
            coinService!!.start()
        }
    }

    override fun onStop() {
        super.onStop()
        coinService!!.stop()
    }

    override fun onDestroy() {
        super.onDestroy()
        disposables!!.dispose()
        currencyRecyclerAdapter = null
        if (activity != null && isBoundCoinService) {
            requireActivity().unbindService(serviceConnection!!)
        }
    }

    companion object {
        @JvmStatic
        fun newInstance(): CurrencyFragment {
            val fragment = CurrencyFragment()
            val args = Bundle()
            // add params if necessary
            // TODO: use navigation safe args
            fragment.arguments = args
            return fragment
        }
    }
}